conn.go

```diff
@@ -604,7 +647,239 @@ func (c *Conn) handleData(arg string) {
 	r.limited = false
 	io.Copy(ioutil.Discard, r) // Make sure all the data has been consumed
 	c.WriteResponse(code, enhancedCode, msg)
+}
+
+// This is a version of io.CopyN with a subtle difference that it returns the number
+// of messages read from src and not the amount of written bytes.
+//
+// This subtle difference is important for use with io.PipeWriter as a destination
+// in handleBdat.
+func copyN(dst io.Writer, src io.Reader, n int64) (consumed int64, err error) {
+	src = io.LimitReader(src, n)
+
+	size := 32 * 1024
+	if l, ok := src.(*io.LimitedReader); ok && int64(size) > l.N {
+		if l.N < 1 {
+			size = 1
+		} else {
+			size = int(l.N)
+		}
+	}
+	buf := make([]byte, size)
+
+	for {
+		nr, er := src.Read(buf)
+		if nr > 0 {
+			consumed += int64(nr)
+			nw, ew := dst.Write(buf[0:nr])
+			if ew != nil {
+				err = ew
+				break
+			}
+			if nr != nw {
+				err = io.ErrShortWrite
+				break
+			}
+		}
+		if er != nil {
+			if er != io.EOF {
+				err = er
+			}
+			break
+		}
+	}
+
+	if consumed < n && err == nil {
+		// src stopped early; must have been EOF.
+		err = io.EOF
+	}
+
+	return consumed, err
+}
+
+func (c *Conn) handleBdat(arg string) {
+	args := strings.Split(arg, " ")
+	if len(args) == 0 {
+		c.WriteResponse(501, EnhancedCode{5, 5, 4}, "Missing chunk size argument")
+		return
+	}
+
+	if !c.fromReceived || len(c.recipients) == 0 {
+		c.WriteResponse(502, EnhancedCode{5, 5, 1}, "Missing RCPT TO command.")
+		return
+	}
+
+	last := false
+	if len(args) == 2 {
```

Comment:
Do we want to be strict or not?

- If we want to be strict, use `strings.SplitN` to avoid having `len(args) > 2`
- If we want to be lax, change this check to `len(args) >= 2` to fail on `BDAT 42 INVALID INVALID`